src/go/monitor/main.go

```diff
@@ -0,0 +1,69 @@
+package main
+
+import (
+	"flag"
+	"log"
+	"net/http"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+// serviceMetrics holds metrics for a service
+type serviceMetrics struct {
+	VisitTotal prometheus.Counter
+}
+
+// newServiceMetrics registers metrics
+func newServiceMetrics() serviceMetrics {
+	return serviceMetrics{
+		VisitTotal: promauto.NewCounter(prometheus.CounterOpts{
+			Name: "myapp_visit_total",
+			Help: "The total number of visits",
+		}),
+	}
+}
+
+// Service of the app
+type Service struct {
+	metrics serviceMetrics
+}
+
+// NewService returns a new service.
+func NewService() Service {
+	return Service{
+		metrics: newServiceMetrics(),
+	}
+}
+
+// Handler return a HTTP handler for the service.
+func (s Service) Handler() http.Handler {
+	mux := http.NewServeMux()
+	mux.Handle("/metrics", promhttp.Handler())
+	return s.middleware(mux)
+}
+
+// middleware to log and record metrics
+func (s Service) middleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		log.Println(r.Method, r.RequestURI)
+		s.metrics.VisitTotal.Inc()
```

Comment:
If this middleware logs and update the metrics on all routes; does it need to be attached to the full `Service`, or can we wrap the server with this middleware at the `ListenAndServe` level instead?

Reply:
It was just a basic middleware as an example. We can remove the logging part. For the monitoring part, my preferences are:
- If we want to monitor the service, keep the method attach to the Service struct
- If we only want to monitor the http handler/router/etc, use the middleware at the ListenAndServe level.